### PR TITLE
Bug 1908217: server-side apply e2e: fix oauthclientauthorization race with KCM

### PR DESCRIPTION
**Pull request by stlaz**
When the oauthclientauthorization is created, it checks the SA from
its clientName for tokens and fails if there are none.

The server-side apply test creates an SA but does not wait for it to
be populated with tokens which may cause the test to fail when
it gets to creating the oauthauthorization object too early.

/assign @marun

### CODE REVIEW

File: test/extended/apiserver/apply.go [hunk 1]
```diff
@@ -172,3 +181,16 @@ func createResource(
...
 }
+
+// waitForSATokens waits for the kube-controller-manager to populate the SA from the unstructured object
+// to be populated with a secret defining its tokens
+func waitForSATokens(saClient corev1client.ServiceAccountInterface, unstructuredObj map[string]interface{}) {
+	name, _, err := unstructured.NestedString(unstructuredObj, "metadata", "name")
+	o.Expect(err).NotTo(o.HaveOccurred())
+	err = wait.PollImmediate(time.Second, 15*time.Second, func() (bool, error) {
+		sa, err := saClient.Get(context.Background(), name, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
```
Comment (marun) [inline test/extended/apiserver/apply.go:192]: I'm surprised to see a hard assertion in a polling loop. Is it your intention that a transient failure fail the test vs logging the error and returning false?

**Review by marun — COMMENTED** [review]
Good catch!

File: test/extended/apiserver/apply.go (hunk 1 above)
Comment (stlaz) [reply test/extended/apiserver/apply.go]: You're right, given the wait is only 15 seconds even if the error does not go away there's no need to fail early.

Fixed now.

File: test/extended/apiserver/apply.go [hunk 2]
```diff
@@ -172,3 +181,19 @@ func createResource(
...
 	return resourceClient, &unstructuredObj, testNamespace
 }
+
+// waitForSATokens waits for the kube-controller-manager to populate the SA from the unstructured object
+// to be populated with a secret defining its tokens
+func waitForSATokens(saClient corev1client.ServiceAccountInterface, unstructuredObj map[string]interface{}) {
+	name, _, err := unstructured.NestedString(unstructuredObj, "metadata", "name")
+	o.Expect(err).NotTo(o.HaveOccurred())
+	err = wait.PollImmediate(time.Second, 15*time.Second, func() (bool, error) {
+		sa, err := saClient.Get(context.Background(), name, metav1.GetOptions{})
```
Comment (sttts) [inline test/extended/apiserver/apply.go:191]: isn't there a variant of PollImmediate with context?

File: test/extended/apiserver/apply.go [hunk 3]
```diff
@@ -172,3 +181,19 @@ func createResource(
 
 	return resourceClient, &unstructuredObj, testNamespace
 }
+
+// waitForSATokens waits for the kube-controller-manager to populate the SA from the unstructured object
+// to be populated with a secret defining its tokens
+func waitForSATokens(saClient corev1client.ServiceAccountInterface, unstructuredObj map[string]interface{}) {
```
Comment (sttts) [inline test/extended/apiserver/apply.go:187]: odd interface. Would be simpler with just the name as string, and the string extracted at the call site.

Comment (stlaz) [reply test/extended/apiserver/apply.go:187]: The function is very specific to this test case, I merely extracted the necessary steps not to pollute the code above.

File: test/extended/apiserver/apply.go (hunk 2 above)
Comment (stlaz) [reply test/extended/apiserver/apply.go:191]: I couldn't find one in `k8s.io/apimachinery/pkg/util/wait/wait.go`

**Pull request merged**